source/coll_db.f90

```diff
@@ -0,0 +1,598 @@
+  character(len=mInputLn) inLine
+  character(len=cdb_fNameLen) cFam
+  real(kind=fPrec)  cNSig
+  integer dbUnit, ioStat, nLines, nSplit, collID, famID, elemEnd
+  integer i, j, ix
+  logical dbErr, spErr
+
+  call f_requestUnit(cdb_fileName, dbUnit)
+  call f_open(unit=dbUnit,file=cdb_fileName,formatted=.true.,mode="r",err=dbErr,status="old")
+  if(dbErr) then
+    write(lout,"(a)") "COLL> ERROR Could not open the collimator database file '"//trim(cdb_fileName)//"'"
+    call prror
+  end if
+
+! ============================================================================ !
+!  Check DB Format: New or old, and count number of collimators
+! ============================================================================ !
+
+  nLines = 0
+
+10 continue
+  read(dbUnit,"(a)",end=20) inLine
+  if(inLine(1:1) == "#") goto 10
+  nLines = nLines + 1
+  if(nLines == 1) then
+    ! Check first line to count number of values
+    call chr_split(inLine, lnSplit, nSplit, spErr)
+    if(nSplit > 1) then
+      cdb_dbOld = .false. ! New style DB (multi-column)
+    else
+      cdb_dbOld = .true.  ! Old style DB (single-column)
+    end if
+  end if
+  goto 10
+
+20 continue
+  call f_close(dbUnit)
+
+  if(cdb_dbOld) then
+    call cdb_readDB_oldFormat
+  else
+    call cdb_readDB_newFormat
+  end if
+
+#ifdef ROOT
+  call cdb_writeDB_ROOT
+#endif
+#ifdef HDF5
+  call cdb_writeDB_HDF5
+#endif
+  call cdb_writeFam
+
+! ============================================================================ !
+!  Post-Processing DB
+! ============================================================================ !
+
+  ! Map single elements to collimators
+  do i=1,iu
+    ix = ic(i)-nblo
+    if(ix < 1) cycle
+
+    collID = -1
+    do j=1,cdb_nColl
+      if(bez(ix) == cdb_cName(j)) then
+        collID = j
+        exit
+      end if
+    end do
+
+    if(collID == -1) then
+      if(bez(ix)(1:2) == "tc" .or. bez(ix)(1:2) == "td" .or. bez(ix)(1:3) == "col") then
+        elemEnd = len_trim(bez(ix))
+        if(bez(ix)(elemEnd-2:elemEnd) /= "_AP") then
+          write(lout,"(a)") "COLLDB> WARNING Collimator not found in database: '"//trim(bez(ix))//"'"
+        end if
+      end if
+    else
+      cdb_elemMap(ix)    = collID
+      cdb_cFound(collID) = .true.
+    end if
+  end do
+
+end subroutine cdb_readCollDB
+
+subroutine cdb_readDB_newFormat
+end subroutine cdb_readDB_newFormat
+
+subroutine cdb_readDB_oldFormat
+
+  use crcoall
+  use parpro
+  use string_tools
+  use mod_units
+  use numerical_constants
+
+  character(len=mInputLn) inLine
+  character(len=cdb_fNameLen) famName
+  logical cErr
+  integer j, dbUnit, dbNew, ioStat, iLine, famID
+
+  cErr = .false.
+
+  call f_requestUnit(cdb_fileName, dbUnit)
+  call f_open(unit=dbUnit,file=cdb_fileName,formatted=.true.,mode="r",status="old")
+
+  call f_requestUnit(trim(cdb_fileName)//".new", dbNew)
+  call f_open(unit=dbNew,file=trim(cdb_fileName)//".new",formatted=.true.,mode="w",status="replace")
```

Comment:
It is done that way for two reasons: Firstly, because this is a help file that will only be generated when the old format is used, and secondly because what is written to that file depends on information generated while doing the loop over file entries. To extract it, I would have to duplicate code.

Reply:
well the only piece of info which is available on the fly is the family name, for which you have a mapping that can be used later, and all the other reported values are from the coll DB, so no duplication of code...
Indeed the new subroutine would be called on purpose only after parsing an old DB - and if you want to debug parsing of the old DB, then I would flush on unit 6